### PR TITLE
Harmonic notch slew limit

### PR DESCRIPTION
**Pull request by andyp1per**
This PR constraints the maximum change allowed per-tick to the harmonic notch center frequency as a percentage of the old frequency. Experiments show that rapid changes to the center frequency can result in filter instability and cascading failures.

This works well and solves the problem on my 5" Beater with KakuteH7v2. Lower values prevent filter instability entirely at some cost to accuracy. Higher values allow more accurate tracking but can result in filter instability in extreme conditions, however this PR also prevents runaway instability.

Flying at 5% (0.05) gave less control in propwash due to the filter tracking less well. Flying at 10% (0.1) made it possible to provoke the issue but in a very subdued form. 7.5% (0.075) turned out to be a good compromise.

### CODE REVIEW

**Comment by andyp1per** [comment]
I flew four batteries on this today and it worked great. Kudos to @lthall for his suggestion as to how to proceed.

**Comment by andyp1per** [comment]
Added to devcall so that @lthall and @tridge can discuss

File: libraries/Filter/NotchFilter.cpp [hunk 1]
```diff
@@ -44,26 +44,36 @@ void NotchFilter<T>::init(float sample_freq_hz, float center_freq_hz, float band
...
+void NotchFilter<T>::init_with_A_and_Q(float sample_freq_hz, float center_freq_hz, float A, float Q, float max_slew)
 {
-    if ((center_freq_hz > 0.0) && (center_freq_hz < 0.5 * sample_freq_hz) && (Q > 0.0)) {
-        float omega = 2.0 * M_PI * center_freq_hz / sample_freq_hz;
+    float new_center_freq = center_freq_hz;
+
+    // constrain the new center frequency by a percentage of the old frequency
+    if (initialised && !need_reset && !is_zero(max_slew)) {
+        new_center_freq = constrain_float(new_center_freq, _center_freq_hz * (1.0 - max_slew),
+                                          _center_freq_hz * (1.0 + max_slew));
```
Comment (lthall) [inline libraries/Filter/NotchFilter.cpp:61]: This should be:
if is_positive(max_slew) {
     new_center_freq = constrain_float(new_center_freq, _center_freq_hz * (1.0 - max_slew),  _center_freq_hz * 1/(1.0 - max_slew));
}

Comment (andyp1per) [reply libraries/Filter/NotchFilter.cpp]: Fixed

**Comment by andyp1per** [comment]
Did extensive testing to find the best value for the slew

**Review by lthall — APPROVED** [review]

**Comment by andyp1per** [comment]
Retested against 4.2 with the latest commit. All good.

**Comment by andyp1per** [comment]
@lthall I did more testing on my 3" and 0.025 is just too low to get good filtering. I had to raise it to 0.05 to get the same noise profile I had before. I think we agreed that anything up to 0.075 was safe, so this feels like a good compromise

**Review by rmackay9 — APPROVED** [review]
LGTM

**Comment by tridge** [comment]
i think there is something missing for RPM sources that update slowly, like EFI RPM and dshot telem then update only happens at low rates, we need to call some update function that will push us towards the desired frequency in between updates

**Comment by tridge** [comment]
as we do the update unconditionally, and I think it is only the apply that is limited to one IMU, then this may cost quite a lot of CPU on something like a CubeBlack

**Pull request merged**